src/topology/config/watcher.rs

```diff
@@ -0,0 +1,127 @@
+use super::Config;
+use futures::{sync::mpsc, Stream};
+use notify::{watcher, DebouncedEvent, RecommendedWatcher, RecursiveMode, Watcher};
+use std::{
+    path::{Path, PathBuf},
+    sync::mpsc::{channel, Receiver},
+    thread,
+    time::Duration,
+};
+
+// Per notify own documentation, it's advised to have delay of more than 30 sec,
+// so to avoid receiving repetitions of previous events on macOS.
+// Larger delays hurt responsivity, but that's fine as this is primarily designed
+// for cloud usage which isn't exactly super responsive.
+pub const CONFIG_WATCH_DELAY: std::time::Duration = std::time::Duration::from_secs(31);
+
+const RETRY_TIMEOUT: std::time::Duration = std::time::Duration::from_secs(10);
+
+/// Returns when config should be reloaded.
+/// Never errors, and never ends.
+pub fn config_watcher(
+    config: &Config,
+    config_path: PathBuf,
+    delay: Duration,
+) -> impl Stream<Item = (), Error = ()> {
+    // Each sender has one slot, and 0 shared slots.
+    let (mut out_sender, out_receiver) = mpsc::channel(0);
+
+    // Fulfills promise of never ending stream.
+    std::mem::forget(out_sender.clone());
```

Comment:
can you explain more why you're doing this with a clone? This seems quite odd

Reply:
Well, it could be added without a clone in else branch of following if statement, but then in the odd case that the watcher thread ever panics it would drop out_sender and break the promise of a never ending stream on which the main function relies on.

That promise could be avoided, but it would require further complicating signaling part of main function or redesigning it.

Note: I am also not a fan of it.

Reply:
... although, it's not by much. So let's try that way.